Dockerfile

```diff
@@ -53,6 +53,11 @@ RUN curl --location https://apt.releases.hashicorp.com/gpg | sudo apt-key add -
  && apt-get clean \
  && rm --recursive --force /var/lib/apt/lists/*
 
+# INSTALL DOCKER
+RUN curl -fsSL https://get.docker.com -o get-docker.sh && sudo sh get-docker.sh && \
+   && apt-get clean \
+   && rm --recursive --force /var/lib/apt/lists/*
+
```

Comment:
Not sure we should do this by default. Maybe leave it to users to do? How much does it increase image size by?

Reply:
not much indeed, we have already discussed to include it.

Reply:
I will check and tell you

Reply:
From 1.76Gb to 2.14Gb @casperdcl

Reply:
Given that we support DinD (Docker in Docker), it would be great to provide it as part of our container images.

Still ~380 MB are a bit too much. 🧐 Are we including the daemon also? If we aren't running it, we should probably install only the command-line tool.

Reply:
This doesn't serve the same purpose as [`docker:dind`](https://hub.docker.com/layers/docker/library/docker/dind/images/sha256-5ed82558a7494a410eb6777aef81cf618118517d352b34307ca40d1da8589a22) and just bloats our images.

If we want to include [`dind`] functionality in our container images, we need to follow the same steps as [`dind/Dockerfile`](https://github.com/docker-library/docker/blob/8baa881aab85f8398d2edbbcc0da4bd1f556dd98/20.10/dind/Dockerfile) and run `dockerd` through an special wrapper at the beginning of the entrypoint.

If we only want to ship the `docker` (client) command–line tool, it should be enough to download [`docker-20.10.9.tgz`](https://download.docker.com/linux/static/stable/x86_64/docker-20.10.9.tgz) from the releases server and install the client binary, 52 MB.

Reply:
@iterative/cml, should we discuss this separately to avoid blocking this pull request?

Reply:
yes 380MB seems much...

```suggestion
RUN apt-get install -yqq apt-transport-https ca-certificates curl \
 && curl -fsSL "https://download.docker.com/linux/ubuntu/gpg" | gpg --dearmor --yes -o /usr/share/keyrings/docker-archive-keyring.gpg \
 && echo "deb [arch=amd64 signed-by=/usr/share/keyrings/docker-archive-keyring.gpg] https://download.docker.com/linux/ubuntu xenial stable" > /etc/apt/sources.list.d/docker.list \
 && apt-get update -qq -o Dir::Etc::sourcelist="sources.list.d/docker.list" -o Dir::Etc::sourceparts="-" -o APT::Get::List-Cleanup="0" \
 && apt-get install -yqq --no-install-recommends docker-ce-cli docker-scan-plugin docker-ce
```

(don't think we need `docker-ce-rootless-extras`)

Reply:
Using the script is what docker recommends to install docker even within docker

Reply:
> _Using the script is what docker recommends to install docker even within docker_
[![wikipedian_protester](https://user-images.githubusercontent.com/11387611/149807803-102f2f82-ff07-480a-8402-cf1e0e897199.png)](https://xkcd.com/285)

Reply:
Resolved in favor of https://github.com/iterative/cml/pull/861/files#discussion_r789262613